Docs: add blog post for C++ 0.3.0 release (#16817)
apache/iceberg: site/docs/blog/posts/2026-06-14-iceberg-cpp-0.3.0-release.md

 -
 - Unless required by applicable law or agreed to in writing, software
 - distributed under the License is distributed on an "AS IS" BASIS,
 - WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 - See the License for the specific language governing permissions and
 - limitations under the License.
 -->

The Apache Iceberg community is pleased to announce the 0.3.0 release of Apache [Iceberg C++](https://github.com/apache/iceberg-cpp). This release includes [over 140 pull requests](https://github.com/apache/iceberg-cpp/compare/v0.2.0...v0.3.0) from 23 contributors, including 11 first-time contributors.

`iceberg-cpp` is a native C++ implementation of the Apache Iceberg table format, providing libraries for reading, writing, and managing Iceberg tables in C++ applications.

<!-- more -->

## Release Highlights

### Scan Planning and Data Access
- [Incremental scan APIs](https://github.com/apache/iceberg-cpp/pull/559), [incremental append scans](https://github.com/apache/iceberg-cpp/pull/590), and [incremental changelog scans](https://github.com/apache/iceberg-cpp/pull/611) for planning table changes between snapshots

> - Merge-on-read data access with a [MOR file scan task reader](https://github.com/apache/iceberg-cpp/pull/657), [delete filter support](https://github.com/apache/iceberg-cpp/pull/650), and a [DeleteLoader for position and equality delete files](https://github.com/apache/iceberg-cpp/pull/610)

Reviewer: ```suggestion
- Merge-on-read data access with a [MOR file scan task reader](https://github.com/apache/iceberg-cpp/pull/657), [delete filter support](https://github.com/apache/iceberg-cpp/pull/650), and a [DeleteLoader for v2 position and equality delete files](https://github.com/apache/iceberg-cpp/pull/610)
```

- [Column selection in table scan planning](https://github.com/apache/iceberg-cpp/pull/550) and [ManifestGroup file filtering](https://github.com/apache/iceberg-cpp/pull/664)
- [Roaring-based position bitmaps](https://github.com/apache/iceberg-cpp/pull/595), a [position delete index](https://github.com/apache/iceberg-cpp/pull/605), and [range coalescing for position deletes](https://github.com/apache/iceberg-cpp/pull/645)

### Table Operations and Maintenance

> - [MergingSnapshotUpdate](https://github.com/apache/iceberg-cpp/pull/682) for rewrite-style operations

Reviewer: ```suggestion
- [MergingSnapshotUpdate](https://github.com/apache/iceberg-cpp/pull/682) lays the groundwork for table overwrite, delete, update, and various maintainances.
```

- [SnapshotManager](https://github.com/apache/iceberg-cpp/pull/542) support and [retried transaction commits](https://github.com/apache/iceberg-cpp/pull/626)
- Snapshot expiration cleanup strategies for [reachable file cleanup](https://github.com/apache/iceberg-cpp/pull/592) and [incremental file cleanup](https://github.com/apache/iceberg-cpp/pull/648)
- [Partition statistics updates](https://github.com/apache/iceberg-cpp/pull/538) and [schema update mapping](https://github.com/apache/iceberg-cpp/pull/561)

### Catalogs and Integrations

> - REST catalog improvements including [initial OAuth2 support](https://github.com/apache/iceberg-cpp/pull/577), [OAuth2 token auto-refresh](https://github.com/apache/iceberg-cpp/pull/646), [Basic authentication](https://github.com/apache/iceberg-cpp/pull/564), [snapshot loading mode](https://github.com/apache/iceberg-cpp/pull/543), [namespace separators](https://github.com/apache/iceberg-cpp/pull/617), and [server-side REST scan planning endpoints](https://github.com/apache/iceberg-cpp/pull/614)

Reviewer: ```suggestion
- REST catalog improvements including [initial OAuth2 support](https://github.com/apache/iceberg-cpp/pull/577), [OAuth2 token auto-refresh](https://github.com/apache/iceberg-cpp/pull/646), [basic authentication](https://github.com/apache/iceberg-cpp/pull/564), [snapshot loading mode](https://github.com/apache/iceberg-cpp/pull/543), [namespace separators](https://github.com/apache/iceberg-cpp/pull/617), and [server-side scan planning endpoints](https://github.com/apache/iceberg-cpp/pull/614)
```

- [S3 FileIO integration](https://github.com/apache/iceberg-cpp/pull/548) built on Arrow filesystem support
- FileIO interface enrichment with new [InputFile and OutputFile interfaces](https://github.com/apache/iceberg-cpp/pull/641) and [bulk delete support](https://github.com/apache/iceberg-cpp/pull/659)
- [SQL catalog support](https://github.com/apache/iceberg-cpp/pull/693) backed by SQLite, PostgreSQL, and MySQL stores

### Metrics and Observability
- Metrics reporter support with [report JSON serialization and reporter loading](https://github.com/apache/iceberg-cpp/pull/589)
- [Avro writer metrics](https://github.com/apache/iceberg-cpp/pull/604) and [Parquet writer metrics](https://github.com/apache/iceberg-cpp/pull/651)

### Metadata and File Format Support
- Puffin support with [basic data structures](https://github.com/apache/iceberg-cpp/pull/588), [format constants and JSON serialization](https://github.com/apache/iceberg-cpp/pull/603), and [file reader/writer support](https://github.com/apache/iceberg-cpp/pull/624)

> - Iceberg v3 support for the [unknown type](https://github.com/apache/iceberg-cpp/pull/662) and [nanosecond timestamp primitive types](https://github.com/apache/iceberg-cpp/pull/653)

Reviewer: ```suggestion
- Iceberg v3 support for the [unknown type](https://github.com/apache/iceberg-cpp/pull/662) and [nanosecond timestamp types](https://github.com/apache/iceberg-cpp/pull/653)
```

 - The ASF licenses this file to You under the Apache License, Version 2.0
 - (the "License"); you may not use this file except in compliance with
 - the License.  You may obtain a copy of the License at
 -
 -   http://www.apache.org/licenses/LICENSE-2.0
 -
 - Unless required by applicable law or agreed to in writing, software
 - distributed under the License is distributed on an "AS IS" BASIS,
 - WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 - See the License for the specific language governing permissions and
 - limitations under the License.
 -->

The Apache Iceberg community is pleased to announce the 0.3.0 release of Apache [Iceberg C++](https://github.com/apache/iceberg-cpp). This release includes [over 140 pull requests](https://github.com/apache/iceberg-cpp/compare/v0.2.0...v0.3.0) from 23 contributors, including 11 first-time contributors.

`iceberg-cpp` is a native C++ implementation of the Apache Iceberg table format, providing libraries for reading, writing, and managing Iceberg tables in C++ applications.

<!-- more -->

## Release Highlights

### Scan Planning and Data Access

> - [Incremental scan APIs](https://github.com/apache/iceberg-cpp/pull/559), [incremental append scans](https://github.com/apache/iceberg-cpp/pull/590), and [incremental changelog scans](https://github.com/apache/iceberg-cpp/pull/611) for planning table changes between snapshots

Reviewer: Should we remove associated URLs? Sometimes a feature involves multiple PRs so it might be confusing to do this. One can always find interesting PRs from https://github.com/apache/iceberg-cpp/releases/tag/v0.3.0. We can just highlight what features have been added in a few sentences.

Author: Yeah, I noticed that the Iceberg C++ 0.2.0 release blog didn't include URLs. But after checking the recent release blogs for Java, Rust, and Go, I saw that they all include relevant URLs, which is why I added them.

I'm not sure if there's a consensus on this. I can remove them if they're not helpful.

- Merge-on-read data access with a [MOR file scan task reader](https://github.com/apache/iceberg-cpp/pull/657), [delete filter support](https://github.com/apache/iceberg-cpp/pull/650), and a [DeleteLoader for position and equality delete files](https://github.com/apache/iceberg-cpp/pull/610)

> - [Column selection in table scan planning](https://github.com/apache/iceberg-cpp/pull/550), [server-side REST scan planning](https://github.com/apache/iceberg-cpp/pull/614), and [ManifestGroup file filtering](https://github.com/apache/iceberg-cpp/pull/664)

Reviewer: I would move `server-side REST scan planning` to section for REST catalog improvements.

- [SnapshotManager](https://github.com/apache/iceberg-cpp/pull/542) support and [retried transaction commits](https://github.com/apache/iceberg-cpp/pull/626)
- Snapshot expiration cleanup strategies for [reachable file cleanup](https://github.com/apache/iceberg-cpp/pull/592) and [incremental file cleanup](https://github.com/apache/iceberg-cpp/pull/648)
- [Partition statistics updates](https://github.com/apache/iceberg-cpp/pull/538) and [schema update mapping](https://github.com/apache/iceberg-cpp/pull/561)

### Catalogs and Integrations
- REST catalog improvements including [initial OAuth2 support](https://github.com/apache/iceberg-cpp/pull/577), [OAuth2 token auto-refresh](https://github.com/apache/iceberg-cpp/pull/646), [Basic authentication](https://github.com/apache/iceberg-cpp/pull/564), [snapshot loading mode](https://github.com/apache/iceberg-cpp/pull/543), [namespace separators](https://github.com/apache/iceberg-cpp/pull/617), and [scan planning endpoints](https://github.com/apache/iceberg-cpp/pull/614)
- [S3 FileIO integration](https://github.com/apache/iceberg-cpp/pull/548) built on Arrow filesystem support

> - [Streaming FileIO APIs](https://github.com/apache/iceberg-cpp/pull/641) and [bulk delete support](https://github.com/apache/iceberg-cpp/pull/659) for more efficient object-store operations

Reviewer: I would say this is `FileIO interface enrichment` because we've introduced InputFile and OutputFile interface to extend its capabilities.

Author: I did some rewording. Please have a look.

- [SQL catalog support](https://github.com/apache/iceberg-cpp/pull/693) backed by SQLite, PostgreSQL, and MySQL stores

> - [Hive metastore build skeleton](https://github.com/apache/iceberg-cpp/pull/689) and [Hive 3.1 metastore Thrift IDL vendoring](https://github.com/apache/iceberg-cpp/pull/694)

Reviewer: ```suggestion
```

I would prefer removing this as it just gets started.

### Metrics and Observability

> - [Scan and commit metrics reporting](https://github.com/apache/iceberg-cpp/pull/589), including report JSON serialization and reporter loading

Reviewer: Let's make it clear that this is just the reporter and metrics and have not been integrated with scan or commit code path.

Author: Reworded so that it describes the reporter and metrics infrastructure only.

- [Avro writer metrics](https://github.com/apache/iceberg-cpp/pull/604) and [Parquet writer metrics](https://github.com/apache/iceberg-cpp/pull/651)

### Metadata and File Format Support
- Puffin support with [basic data structures](https://github.com/apache/iceberg-cpp/pull/588), [format constants and JSON serialization](https://github.com/apache/iceberg-cpp/pull/603), and [file reader/writer support](https://github.com/apache/iceberg-cpp/pull/624)
- Iceberg v3 support for the [unknown type](https://github.com/apache/iceberg-cpp/pull/662) and [nanosecond timestamp primitive types](https://github.com/apache/iceberg-cpp/pull/653)
- Expression serialization with [operation JSON serialization](https://github.com/apache/iceberg-cpp/pull/532), [expression JSON serialization](https://github.com/apache/iceberg-cpp/pull/553), and [typed literal binding after serialization](https://github.com/apache/iceberg-cpp/pull/562)

> - [Secure UUID generation](https://github.com/apache/iceberg-cpp/pull/602) and [table UUID assignment during metadata creation](https://github.com/apache/iceberg-cpp/pull/713)

Reviewer: I think these two are just bug fixes.

Author: removed together with the below section.

> ### Build, Packaging, and Documentation

Reviewer: Perhaps we don't need to mention these items in a release blog because they are not **features**.
